tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java

```diff
@@ -127,163 +128,207 @@ public void resolve() {
-          )
-      );
-      executorBuilder.clear();
-    }
+      indexScanBuilder
+          .setTableId(tableInfo.getId())
+          .setIndexId(indexInfo.getId());
+      dagRequestBuilder.addExecutors(executorBuilder.setIdxScan(indexScanBuilder).build());
+    } else {
+      // TableScan
+      executorBuilder.setTp(ExecType.TypeTableScan);
+      tblScanBuilder.setTableId(tableInfo.getId());
+      // Step1. Add columns to first executor
+      for (int i = 0; i < getFields().size(); i++) {
+        ColumnRef col = getFields().get(i);
+        tblScanBuilder.addColumns(col.getColumnInfo().toProto(tableInfo));
+        colOffsetMap.put(col, i);
+      }
+      // Currently, according to TiKV's implementation, if handle
+      // is needed, we should add an extra column with an ID of -1
+      // to the TableScan executor
+      if (isHandleNeeded()) {
+        ColumnInfo handleColumn = ColumnInfo.newBuilder()
+            .setColumnId(-1)
+            .setPkHandle(true)
+            // We haven't changed the field name in protobuf file, but
+            // we need to set this to true in order to retrieve the handle,
+            // so the name 'setPkHandle' may sounds strange.
+            .build();
+        tblScanBuilder.addColumns(handleColumn);
+      }
+      dagRequestBuilder.addExecutors(executorBuilder.setTblScan(tblScanBuilder));
 
-    if (!getGroupByItems().isEmpty() || !getAggregates().isEmpty()) {
-      Aggregation.Builder aggregationBuilder = Aggregation.newBuilder();
-      getGroupByItems().forEach(tiByItem -> aggregationBuilder.addGroupBy(ProtoConverter.toProto(tiByItem.getExpr(), colOffsetMap)));
-      getAggregates().forEach(tiExpr -> aggregationBuilder.addAggFunc(ProtoConverter.toProto(tiExpr, colOffsetMap)));
-      executorBuilder.setTp(ExecType.TypeAggregation);
-      dagRequestBuilder.addExecutors(
-          executorBuilder.setAggregation(aggregationBuilder)
-      );
-      executorBuilder.clear();
+      // column offset should be in accordance with fields
+      for (int i = 0; i < getFields().size(); i ++) {
+        dagRequestBuilder.addOutputOffsets(i);
+      }
+
+      // if handle is needed, we should append one output offset
+      if (isHandleNeeded()) {
+        dagRequestBuilder.addOutputOffsets(tableInfo.getColumns().size());
+      }
     }
 
-    if (!getOrderByItems().isEmpty()) {
-      TopN.Builder topNBuilder = TopN.newBuilder();
-      getOrderByItems().forEach(tiByItem -> topNBuilder
-          .addOrderBy(com.pingcap.tidb.tipb.ByItem.newBuilder()
-          .setExpr(ProtoConverter.toProto(tiByItem.getExpr(), colOffsetMap))
-          .setDesc(tiByItem.isDesc())
-      ));
-      executorBuilder.setTp(ExecType.TypeTopN);
-      topNBuilder.setLimit(getLimit());
-      dagRequestBuilder.addExecutors(executorBuilder.setTopN(topNBuilder));
+    if (!isIndexScan || (isIndexScan() && !isDoubleRead())) {
```

Comment:
For table read and index single read:
Filters, OrderBy, groupBy and limit should be added to dagRequest.